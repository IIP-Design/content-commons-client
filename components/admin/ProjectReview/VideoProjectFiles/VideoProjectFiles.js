/**
 *
 * VideoProjectFiles
 *
 */
import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import { graphql } from 'react-apollo';
import { Button, Grid, Loader } from 'semantic-ui-react';

import ApolloError from 'components/errors/ApolloError';
import VideoProjectFile from './VideoProjectFile/VideoProjectFile';
import { getPluralStringOrNot } from 'lib/utils';
import { VIDEO_PROJECT_PREVIEW_QUERY } from 'components/admin/Previews/ProjectPreview/ProjectPreviewContent/ProjectPreviewContent';

import './VideoProjectFiles.scss';


const VideoProjectFiles = ( { data, id } ) => {
  const { error, loading, project } = data;

  if ( loading ) {
    return (
      <div
        style={ {
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          minHeight: '200px',
        } }
      >
        <Loader
          active
          inline="centered"
          style={ { marginBottom: '1em' } }
          content="Loading project file(s)..."
        />
      </div>
    );
  }

  if ( error ) return <ApolloError error={ error } />;
  if ( !project || !Object.keys( project ).length ) return null;

  const { units } = project;

  if ( !units || units.length === 0 ) return null;

  const getTag = ( tag, unit ) => {
    const translation = tag.translations.find( t => t.language.locale === unit.language.locale );

    if ( translation && translation.name ) {
      return translation.name;
    }
  };

  const getTags = ( tags, unit ) => tags.reduce( ( acc, curr ) => `${acc ? `${acc}, ` : ''}${getTag( curr, unit )}`
    , '' );
  const handleEdit = () => {
    Router.push( {
      pathname: '/admin/project',
      query: {
        id,
        content: 'video',
        action: 'edit',
      },
    }, `/admin/project/video/${id}/edit` );
  };

  return (
    <section className="section section--project_units project_units layout">
      <h3 className="project_units_headline uppercase">
        { `${getPluralStringOrNot( units, 'Video' )} in Project` }
      </h3>
      { units.map( unit => {
        const {
          files, language, tags, thumbnails,
        } = unit;

        return (
          <div key={ unit.id } className="project_unit">
            <Grid>
              <Grid.Row
                className={ `project_unit_header ${unit.language.textDirection}` }
              >
                <Grid.Column floated="left" mobile={ 8 }>
                  <h4 className="title">{ unit.title }</h4>
                </Grid.Column>
                <Grid.Column floated="right" mobile={ 8 } className="project_unit_edit">
                  <Button
                    className="project_button project_button--edit"
                    content="Edit"
                    onClick={ handleEdit }
                  />
                </Grid.Column>
              </Grid.Row>

              <Grid.Row className="project_unit_meta language">
                <Grid.Column width={ 16 }>
                  <p>
                    <b className="label">Language:</b>
                    { ` ${language.displayName} `}
                  </p>
                </Grid.Column>
              </Grid.Row>

              <Grid.Row className="project_unit_meta description">
                <Grid.Column computer={ 8 }>
                  <p className="public_description">
                    <b className="label">PublicÂ Description:</b>
                    <span
                      className={ `content ${language.textDirection}` }
                      style={
                        language.textDirection === 'RTL'
                          ? { display: 'block' }
                          : {}
                      }
                    >
                      { ` ${unit.descPublic || project.descPublic} ` }
                    </span>
                  </p>
                </Grid.Column>
              </Grid.Row>

              <Grid.Row className="project_unit_meta tags">
                <Grid.Column mobile={ 16 } tablet={ 8 } computer={ 8 }>
                  <p>
                    <b className="label">Additional Keywords:</b>
                    { ` ${tags && tags.length > 0
                      ? getTags( tags, unit )
                      : ''}` }
                  </p>
                </Grid.Column>
              </Grid.Row>

              <Grid.Row className="project_unit_meta file-count">
                <Grid.Column width={ 16 }>
                  <p>
                    <b className="label">Files:</b>
                    { ` ${files?.length || 0} ` }
                  </p>
                </Grid.Column>
              </Grid.Row>

              { files && files.length > 0 && files.map( file => (
                <VideoProjectFile
                  key={ file.id }
                  file={ file }
                  thumbnail={ thumbnails && thumbnails.length > 0 && thumbnails[0] }
                />
              ) ) }
            </Grid>
          </div>
        );
      } ) }
    </section>
  );
};

VideoProjectFiles.propTypes = {
  id: PropTypes.string,
  data: PropTypes.object,
};

export default graphql( VIDEO_PROJECT_PREVIEW_QUERY, {
  options: props => ( {
    variables: { id: props.id },
  } ),
} )( VideoProjectFiles );
