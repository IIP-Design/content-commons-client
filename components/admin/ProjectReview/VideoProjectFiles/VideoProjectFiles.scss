@import "styles/colors.scss";
@import "styles/mixins.scss";

.project_units {
  &_headline {
    @media screen and (max-width: 767px) {
      margin-right: initial;
      margin-left: initial;
    }
    margin-right: -1rem;
    margin-left: -1rem;
    padding: 1rem;
    border-top: 2px solid $light-grey;
    border-bottom: 2px solid $light-grey;
  }
}

.project_unit {
  margin-top: 1em;
  @media screen and (min-width: 768px) {
    padding-bottom: 1em;
  }
  border-bottom: 2px solid $light-grey;

  > .ui.grid {
    > .row {
      @media screen and (max-width: 767px) {
        padding-left: 1rem;
        padding-right: 1rem;
      }

      &[class$="_files"] {
        padding-top: 0.5rem;
      }
    }
  }

  &_header {
    padding-bottom: 0 !important;

    &.RTL {
      .title {
        direction: rtl;
        text-align: right;
      }
    }
  }

  &_meta {
    padding-top: 0 !important;
    @media screen and (min-width: 768px) {
      padding-bottom: 0.5rem !important;
    }

    &.description .content {
      font-size: 0.777777778rem;
      &.RTL {
        direction: rtl;
        text-align: right;
      }
    }

    &.tags p {
      @include wrap-text;
    }
  }

  &_edit {
    text-align: right;
  }

  .file_meta,
  .file_embed_thumbnail {
    @media screen and (max-width: 767px) {
      padding-bottom: 1rem;
    }

    img {
      height: auto;
      max-width: 100%;
      width: 100%;
    }
  }
}
