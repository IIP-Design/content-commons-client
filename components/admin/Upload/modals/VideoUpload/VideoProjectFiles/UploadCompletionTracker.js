import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Label, Icon } from 'semantic-ui-react';

/**
 * Tracks number of fields that need to be completed
 * Props:
 * fields: Fields to track completion
 * display: Should completion tracker be shown in parent component
 *
 * @param {object } props
 *
 */
const UploadCompletionTracker = props => {
  const { fields, display } = props;
  const [fieldsToComplete, setFieldsToComplete] = useState( fields.length );

  // fields are properties of a file object that is stored in main files state array
  // in the VideoUpload component so when another component changes the field,
  // it will be reflected here. Field completion is marked when a field has a
  // valid value
  useEffect( () => {
    const toComplete = fields.filter( field => !field );
    setFieldsToComplete( toComplete.length );
  }, [fields] );


  // following styles used to display either completion OR number of fields to complete
  const show = {
    display: 'inline-block'
  };

  const hide = {
    display: 'none'
  };

  return (
    <span className={ `item-complete ${display}` }>
      <Label style={ fieldsToComplete ? show : hide } circular size="mini" color="red">{ fieldsToComplete }</Label>
      <Icon style={ fieldsToComplete ? hide : show } circular size="small" inverted color="green" name="check" />
    </span>
  );
};

UploadCompletionTracker.defaultProps = {
  display: 'show'
};

UploadCompletionTracker.propTypes = {
  fields: PropTypes.array,
  display: PropTypes.string
};

export default UploadCompletionTracker;
