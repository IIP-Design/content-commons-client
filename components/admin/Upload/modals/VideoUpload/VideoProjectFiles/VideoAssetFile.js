import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Grid, Icon } from 'semantic-ui-react';
import replaceIcon from 'static/icons/icon_replace.svg';
import removeIcon from 'static/icons/icon_remove.svg';
import 'styles/tooltip.scss';
import { truncateAndReplaceStr } from 'lib/utils';
import { isWindowWidthLessThanOrEqualTo } from 'lib/browser';
import VideoAssetOptions from '../VideoAssetOptions/VideoAssetOptions';
import './VideoAssetFile.scss';

class VideoAssetFile extends PureComponent {
  state = {
    isMobileDevice: false,
    actionsMenuActive: false,
    selectOptionsActive: false,
    language: null,
    subtitles: null,
    typeUse: null,
    quality: null,
    optionsCount: 0,
  }

  componentDidMount() {
    const isMobileDevice = isWindowWidthLessThanOrEqualTo( 640 );

    window.addEventListener( 'resize', this.fileOnResize );

    const options = Array.from( this.fileAsset.querySelectorAll( '.videoProjectFiles_asset_options' ) );
    const subtitles = options.filter( option => option.classList.contains( 'subtitles' ) )[0];
    const subtitlesCount = subtitles.className.includes( 'notApplicable' ) ? 0 : 1;
    const languageCount = 1;

    this.setState( {
      isMobileDevice,
      optionsCount: languageCount + subtitlesCount
    } );
  }

  componentDidUpdate() {
    const {
      isMobileDevice,
      language,
      subtitles,
      typeUse,
      quality
    } = this.state;
    const options = Array.from( this.fileAsset.querySelectorAll( '.videoProjectFiles_asset_options' ) );
    const languageCount = language !== null ? 0 : 1;

    let typeUseDOM;
    let typeUseCount;
    let qualityDOM;
    let qualityCount;

    if ( isMobileDevice ) {
      const subtitlesDOM = options.filter( option => option.classList.contains( 'subtitles' ) )[0];
      const subtitlesCount = subtitlesDOM.className.includes( 'notApplicable' ) || subtitles !== null
        ? 0
        : 1;
      [typeUseDOM] = options.filter( option => option.classList.contains( 'typeUse' ) );
      typeUseCount = typeUseDOM.className.includes( 'notApplicable' ) || typeUse !== null
        ? 0
        : 1;
      [qualityDOM] = options.filter( option => option.classList.contains( 'quality' ) );
      qualityCount = qualityDOM.className.includes( 'notApplicable' ) || quality !== null
        ? 0
        : 1;
      /* eslint-disable-next-line react/no-did-update-set-state */
      return this.setState( {
        optionsCount: languageCount + subtitlesCount + typeUseCount + qualityCount
      } );
    }

    if ( this.props.activeStep === 'step_1' ) {
      const subtitlesDOM = options.filter( option => option.classList.contains( 'subtitles' ) )[0];
      const subtitlesCount = subtitlesDOM.className.includes( 'notApplicable' ) || subtitles !== null
        ? 0
        : 1;
      /* eslint-disable-next-line react/no-did-update-set-state */
      return this.setState( { optionsCount: languageCount + subtitlesCount } );
    }

    [typeUseDOM] = options.filter( option => option.classList.contains( 'typeUse' ) );
    typeUseCount = typeUseDOM.className.includes( 'notApplicable' ) || typeUse !== null
      ? 0
      : 1;
    [qualityDOM] = options.filter( option => option.classList.contains( 'quality' ) );
    qualityCount = qualityDOM.className.includes( 'notApplicable' ) || quality !== null
      ? 0
      : 1;
    /* eslint-disable-next-line react/no-did-update-set-state */
    return this.setState( { optionsCount: typeUseCount + qualityCount } );
  }

  componentWillUnmount() {
    window.removeEventListener( 'resize', this.fileOnResize );
  }

  handleLanguageChange = language => {
    this.props.setOptionsComplete();
    this.setState( { language } );
  }

  handleSubtitlesChange = subtitles => {
    this.props.setOptionsComplete();
    this.setState( { subtitles } );
  }

  handleTypeUseChange = typeUse => {
    this.props.setOptionsComplete();
    this.setState( { typeUse } );
  }

  handleQualityChange = quality => {
    this.props.setOptionsComplete();
    this.setState( { quality } );
  }

  fileOnResize = () => {
    let fileResizeTimer = null;
    clearTimeout( fileResizeTimer );
    fileResizeTimer = setTimeout( () => {
      const mobileWidth = isWindowWidthLessThanOrEqualTo( 640 );
      this.setState( { isMobileDevice: mobileWidth } );
    }, 500 );
  }

  toggleAssetFileActionsMenu = () => this.setState( prevState => ( {
    actionsMenuActive: !prevState.actionsMenuActive
  } ) );

  toggleSelectOptionsDisplay = () => this.setState( prevState => ( {
    selectOptionsActive: !prevState.selectOptionsActive
  } ) );

  render() {
    const {
      activeStep,
      file,
      removeVideoAssetFile,
      replaceVideoAssetFile
    } = this.props;

    const {
      isMobileDevice,
      actionsMenuActive,
      selectOptionsActive,
      language,
      subtitles,
      typeUse,
      quality,
      optionsCount,
    } = this.state;

    let fileDisplayName;
    if ( file.length > 30 ) {
      fileDisplayName = isMobileDevice
        ? truncateAndReplaceStr( file, 10, 9 )
        : truncateAndReplaceStr( file, 18, 9 );
    } else {
      fileDisplayName = file;
    }

    return (
      <Grid.Row className="videoProjectFiles_asset">
        <Grid.Column mobile={ 16 } tablet={ 6 } computer={ 6 } className="videoProjectFiles_asset_file_wrapper">
          <p
            className={
              optionsCount > 0
                ? 'videoProjectFiles_asset_file incomplete'
                : 'videoProjectFiles_asset_file complete'
            }
            title={ file }
            data-optionscount={ optionsCount }
          >
            { fileDisplayName }
          </p>
          <p className={ `videoProjectFiles_asset_file--fullFileName ${selectOptionsActive ? 'active' : ''}` }>
            <Icon name="info" size="small" />
            { file }
          </p>
        </Grid.Column>
        <Grid.Column mobile={ 16 } tablet={ 10 } computer={ 10 }>
          <div
            ref={ fileAsset => { this.fileAsset = fileAsset; } }
            className={ `videoProjectFiles_asset_options_wrapper ${selectOptionsActive ? 'active' : ''}` }
          >
            <VideoAssetOptions
              isMobileDevice={ isMobileDevice }
              activeStep={ activeStep }
              file={ file }
              language={ language }
              handleLanguageChange={ this.handleLanguageChange }
              subtitles={ subtitles }
              handleSubtitlesChange={ this.handleSubtitlesChange }
              typeUse={ typeUse }
              handleTypeUseChange={ this.handleTypeUseChange }
              quality={ quality }
              handleQualityChange={ this.handleQualityChange }
            />
          </div>
          { !isMobileDevice && (
            <div className="videoProjectFiles_asset_actionBtns">
              <span tooltip="Replace">
                <label htmlFor="replace_videoAssetFile" className="videoProjectFiles_asset_replaceBtn">
                  <input
                    type="file"
                    name="replace_videoAssetFile"
                    data-filename={ file }
                    ref={ ref => { this.replaceFile = ref; } }
                    onChange={ replaceVideoAssetFile }
                  />
                  <button type="button" onClick={ () => this.replaceFile.click() }>
                    <img src={ replaceIcon } alt="Replace Video File Button" />
                  </button>
                </label>
              </span>
              <span tooltip="Remove">
                <button type="button" onClick={ () => removeVideoAssetFile( file ) }>
                  <img src={ removeIcon } alt="Remove Video File Button" className="videoProjectFiles_asset_removeBtn" />
                </button>
              </span>
              <span
                className="actionsMenuCloseButton"
                role="button"
                onClick={ this.toggleAssetFileActionsMenu }
                onKeyPress={ this.toggleAssetFileActionsMenu }
                tabIndex="0"
              />
            </div>
          ) }
        </Grid.Column>

        { /* MOBILE ELEMENTS */ }
        <Icon
          name={ `${selectOptionsActive ? 'chevron up' : 'chevron down'}` }
          size="small"
          className="mobileSelectOptionsToggle"
          onClick={ this.toggleSelectOptionsDisplay }
          onKeyPress={ this.toggleSelectOptionsDisplay }
        />
        { !actionsMenuActive && (
          <Icon
            name="ellipsis vertical"
            size="small"
            className="mobileActionsMenuToggle"
            onClick={ this.toggleAssetFileActionsMenu }
            onKeyPress={ this.toggleAssetFileActionsMenu }
          />
        ) }
        { actionsMenuActive && isMobileDevice && (
          <div className="videoProjectFiles_asset_actionBtns active">
            <label htmlFor="replace_videoAssetFile" className="videoProjectFiles_asset_replaceBtn">
              <input
                type="file"
                name="replace_videoAssetFile"
                data-filename={ file }
                ref={ ref => { this.replaceFile = ref; } }
                onChange={ replaceVideoAssetFile }
              />
              <button type="button" onClick={ () => this.replaceFile.click() }>
                <span>Replace File</span>
              </button>
            </label>
            <button
              className="videoProjectFiles_asset_removeBtn"
              type="button"
              onClick={ () => removeVideoAssetFile( file ) }
            >
              <span>Delete File</span>
            </button>
            <span
              className="actionsMenuCloseButton"
              role="button"
              onClick={ this.toggleAssetFileActionsMenu }
              onKeyPress={ this.toggleAssetFileActionsMenu }
              tabIndex="0"
            />
          </div>
        ) }
      </Grid.Row>
    );
  }
}

VideoAssetFile.propTypes = {
  file: PropTypes.string,
  activeStep: PropTypes.string,
  removeVideoAssetFile: PropTypes.func,
  replaceVideoAssetFile: PropTypes.func,
  setOptionsComplete: PropTypes.func
};

export default VideoAssetFile;
