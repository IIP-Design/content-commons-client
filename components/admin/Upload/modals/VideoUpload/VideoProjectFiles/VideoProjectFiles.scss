@import "styles/colors.scss";

.ui.modal.upload_modal {

  &.scrolling {
    @media screen and (max-width: 640px) {
      overflow-x: hidden;
    }
  }  

  &.prepare-files-active {
    max-width: 788px;
  }  
}

.videoProjectFiles {    

  &.ui.form {
    max-width: 640px;
    margin: 0 auto;

    > .ui.grid {
      @media screen and (min-width: 641px) and (max-width: 900px) {
        padding-right: 1em;
        padding-left: 1em;
      }
    }
  }

  &_headline {
    margin-top: -45px !important;
  }

  .ui.grid {
    margin: 0;

    .row {
      padding-top: 0.5em;
      padding-bottom: 0;

      > .column.six {
        padding-left: 0;
        @media screen and (max-width: 640px) {
          padding: 0;
        }
        @media only screen and (min-width: 641px) and (max-width: 991px) {
          width: 37.5% !important;
        }
      }      

      > .column.ten {
        position: relative;
        padding-right: 0;
        @media screen and (max-width: 690px) {
          padding: 0;
        }
        @media only screen and (min-width: 641px) and (max-width: 991px) {
          width: 62.5% !important;
        }
      }
    }
  }

  &_column_labels {
    padding-bottom: 0.5em !important;
    border-bottom: 1px solid $light-grey;
    color: $blue;
    @media screen and (max-width: 640px) {
      display: none;
    }
  }

  &_column_label {
    display: inline-block;
    width: 148px;
    margin-bottom: 0;
    @media screen and (max-width: 640px) {
      display: none;
    }
    @media screen and (min-width: 641px) and (max-width: 900px) {
      width: 125px;
    }    

    &:first-child {
      margin-right: 16px;
    }

    &--required {
      position: relative;

      &:after {
        content: '\2605';
        position: absolute;
        top: -3px;
        font-size: 6px;
        color: red;
      }
    }    
  }

  &_steps {
    text-align: center;
    @media screen and (max-width: 640px) {
      display: none;
    }

    > p {
      display: inline-block;
    }

    > i {
      margin: 0 2em;
      vertical-align: middle;
    }
  }

  &_step {
    font-size: 20px;
    color: $alt-grey;

    &.active {
      position: relative;
      font-weight: bold;
      color: $blue;

      &:after {
        content: '';
        position: absolute;
        left: 0;
        bottom: -5px;
        width: 125%;
        height: 5px;
        background-color: $light-blue;
        transform: translateX(-10%);
      }
    }        
  }    
}