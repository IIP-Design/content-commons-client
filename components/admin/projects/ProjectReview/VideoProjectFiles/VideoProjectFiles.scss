@import "../../../../../styles/colors.scss";

.project_files {

  &_headline {
    position: relative;
    padding-left: 1rem;
    padding-top: 1rem;
    padding-bottom: 1rem;
    margin-right: -1rem;
    margin-left: -1rem;

    &:after,
    &:before {
      content: "";
      position: absolute;
      left: 0;  
      height: 2px;
      width: 100%;
      background-color: $light-grey;
    }

    &:after {
      bottom: 0;
    }

    &:before {
      top: 0;
    }
  }
}

 
.review-project .project_file {
  position: relative;  
  margin-top: 1em;
  @media screen and (min-width: 768px) {    
    padding-bottom: 2em;
  }

  &:after {
    content: "";
    position: absolute;
    bottom: 0;
    left: 0;  
    height: 2px;
    width: 100%;
    background-color: $light-grey;
  } 

  &_header {
    padding-bottom: 0 !important;

    &.RTL {

      .title {
        direction: rtl;
      }      
    }
  }

  &_contents {
    @media screen and (min-width: 768px) {
      padding-top: 0 !important;
    }        
  }

  &_edit {
    text-align: right;
  }

  .file_meta {
    padding-right: 0;

    img {
      width: 100%;
      margin-bottom: 0.5em;
    }        
  }

  .file_info {

    .public_description,
    .additional_keywords {
      margin: 1em 0;
    }

    .public_description {

      span {
        display: block;
      }

      &.RTL {

        span:last-child {
          direction: rtl;
        }        
      }
    }    
  }

}