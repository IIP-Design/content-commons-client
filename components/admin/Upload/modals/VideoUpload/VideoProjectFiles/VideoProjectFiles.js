import React, { useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import VideoProjectFilesDesktop from './VideoProjectFileDesktop/VideoProjectFileDesktop';
import VideoProjectFilesMobile from './VideoProjectFilesMobile/VideoProjectFilesMobile';
import { VideoUploadContext } from '../VideoUpload';
import './VideoProjectFiles.scss';

const VideoProjectFiles = props => {
  // since we are using a stateless function, use a hook for mouunting/unmouting calls
  useEffect( () => {
    props.updateModalClassname( 'upload_modal prepare-files-active' );
    return () => {
      props.updateModalClassname( 'upload_modal' );
    };
  }, [] );


  return (
    <VideoUploadContext.Consumer>
      { context => {
        const message = context.files.length
          ? `Preparing ${context.files.length} files for upload...`
          : 'Please add files to upload';

        return (
          <Fragment>
            <h5 className="videoProjectFiles__headline">{ message }</h5>
            <VideoProjectFilesDesktop />
            <VideoProjectFilesMobile />
          </Fragment>
        );
      } }
    </VideoUploadContext.Consumer>
  );
};


VideoProjectFiles.propTypes = {
  updateModalClassname: PropTypes.func
};

export default VideoProjectFiles;
