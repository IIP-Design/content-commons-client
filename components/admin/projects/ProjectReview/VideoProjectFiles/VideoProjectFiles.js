/**
 *
 * VideoProjectFiles
 *
 */

import React from 'react';
import { object, string } from 'prop-types';
import Router from 'next/router';
import { graphql } from 'react-apollo';
import { Grid, Button } from 'semantic-ui-react';

import { VIDEO_PROJECT_PREVIEW_QUERY } from 'components/admin/projects/ProjectEdit/PreviewProjectContent/PreviewProjectContent';

import {
  formatBytes,
  formatDate,
  getPluralStringOrNot,
  millisToMinutesAndSeconds
} from 'lib/utils';

import './VideoProjectFiles.scss';

const VideoProjectFiles = props => {
  const { error, loading, project } = props.data;

  if ( loading ) return 'Loading the project...';
  if ( error ) return `Error! ${error.message}`;

  if ( !project || !Object.keys( project ).length ) return null;

  const { units } = project;

  const handleEdit = () => {
    const { id } = props;
    Router.push( {
      pathname: `/admin/project/unit/${id}/edit`
    } );
  };

  return (
    <section className="section section--project_files project_files layout">
      <h3 className="project_files_headline uppercase">
        { `${getPluralStringOrNot( units, 'Video' )} in Project` }
      </h3>
      { units.map( unit => (
        <div key={ unit.title } className="project_file">
          <Grid>
            <Grid.Row
              className={
                `project_file_header ${unit.files[0].language.textDirection}`
              }
            >
              <Grid.Column floated="left" mobile={ 8 }>
                <h4 className="title">{ unit.title }</h4>
              </Grid.Column>
              <Grid.Column floated="right" mobile={ 8 } className="project_file_edit">
                <Button
                  className="project_button project_button--edit"
                  content="Edit"
                  onClick={ handleEdit }
                />
              </Grid.Column>
            </Grid.Row>

            <Grid.Row className="project_file_contents">
              <Grid.Column mobile={ 16 } tablet={ 8 } computer={ 8 } className="file_meta">
                <img src={ unit.thumbnails[0].image.url } alt={ unit.thumbnails[0].image.alt } />
                <p><b className="label">File Name:</b> { unit.files[0].filename }</p>
                <p><b className="label">Filesize:</b> { formatBytes( unit.files[0].filesize ) }</p>
                <p><b className="label">Dimensions:</b> { `${unit.files[0].dimensions.width} x ${unit.files[0].dimensions.height}` }</p>
                <p><b className="label">Uploaded:</b> { formatDate( unit.files[0].createdAt ) }</p>
                <p><b className="label">Duration:</b> { millisToMinutesAndSeconds( unit.files[0].duration, 0 ) }</p>
              </Grid.Column>

              <Grid.Column mobile={ 16 } tablet={ 8 } computer={ 8 } className="file_info">
                <p><b className="label">Language:</b> { unit.files[0].language.displayName }</p>
                <p><b className="label">Subtitles & Captions:</b> { unit.files[0].videoBurnedInStatus }</p>
                <p><b className="label">Video Type:</b> { unit.videoType }</p>
                <p><b className="label">Quality:</b> { unit.files[0].quality }</p>
                <p
                  className={
                    `public_description ${unit.files[0].language.textDirection}`
                  }
                >
                  <b className="label">PublicÂ Description:</b>
                  <span>{ unit.descPublic }</span>
                </p>
                <p><b className="label">{ unit.files[0].stream.site } URL:</b> { unit.files[0].stream.url }</p>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        </div>
      ) ) }
    </section>
  );
};

VideoProjectFiles.propTypes = {
  id: string,
  data: object
};

export default graphql( VIDEO_PROJECT_PREVIEW_QUERY, {
  options: props => ( {
    variables: {
      id: props.id,
      isReviewPage: true
    },
  } )
} )( VideoProjectFiles );
