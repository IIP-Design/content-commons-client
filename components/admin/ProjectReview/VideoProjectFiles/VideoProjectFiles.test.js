import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Router from 'next/router';
import { MockedProvider } from '@apollo/react-testing';
import { Loader } from 'semantic-ui-react';

import VideoProjectFiles from './VideoProjectFiles';

import {
  errorMocks,
  mocks,
  noFilesMocks,
  noUnitsMocks,
  nullFilesMocks,
  nullMocks,
  nullUnitsMocks,
  props,
} from 'components/admin/Previews/ProjectPreview/ProjectPreviewContent/mocks';

jest.mock( 'next/config', () => () => ( { publicRuntimeConfig: { REACT_APP_AWS_S3_AUTHORING_BUCKET: 's3-bucket-url' } } ) );

jest.mock( 'lib/utils', () => ( {
  getPluralStringOrNot: jest.fn( ( array, string ) => `${string}${array && array.length > 1 ? 's' : ''}` ),
  getApolloErrors: jest.fn( error => {
    let errs = [];
    const { graphQLErrors, networkError, otherError } = error;

    if ( graphQLErrors ) {
      errs = graphQLErrors.map( error => error.message );
    }
    if ( networkError ) errs.push( networkError );
    if ( otherError ) errs.push( otherError );

    return errs;
  } ),
} ) );

jest.mock( './VideoProjectFile/VideoProjectFile', () => 'video-project-file' );

const Component = (
  <MockedProvider mocks={ mocks } addTypename>
    <VideoProjectFiles { ...props } />
  </MockedProvider>
);

const suppressActWarning = consoleError => {
  const actMsg = 'Warning: An update to %s inside a test was not wrapped in act';

  jest.spyOn( console, 'error' ).mockImplementation( ( ...args ) => {
    if ( !args[0].includes( actMsg ) ) {
      consoleError( ...args );
    }
  } );
};

describe( '<VideoProjectFiles />', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
        content="Loading project file(s)..."
      />
    );

    expect( videoProjectFiles.exists() ).toEqual( true );
    expect( videoProjectFiles.contains( loader ) ).toEqual( true );
  } );

  it( 'renders error message if an error is thrown', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename>
        <VideoProjectFiles { ...props } />
      </MockedProvider>,
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );
    const errorComponent = videoProjectFiles.find( 'ApolloError' );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders `null` if project is `null`', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename>
        <VideoProjectFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );

    expect( videoProjectFiles.html() ).toEqual( null );
  } );

  it( 'renders the final state', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );

    expect( toJSON( videoProjectFiles ) ).toMatchSnapshot();
  } );

  it( 'clicking the Edit button redirects to <VideoEdit />', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );
    const header = videoProjectFiles.find( '.project_file_header' );
    const editBtns = header.find( 'Button.project_button--edit' );

    Router.push = jest.fn();

    editBtns.forEach( btn => {
      const { id } = props;

      btn.simulate( 'click' );
      expect( Router.push ).toHaveBeenCalledWith( {
        pathname: '/admin/project',
        query: {
          id,
          content: 'video',
          action: 'edit',
        },
      }, `/admin/project/video/${id}/edit` );
    } );
  } );

  it( 'renders `null` if units is `null`', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullUnitsMocks } addTypename>
        <VideoProjectFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );

    expect( videoProjectFiles.exists() ).toEqual( true );
    expect( videoProjectFiles.html() ).toEqual( null );
  } );

  it( 'renders `null` if units is `[]`', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noUnitsMocks } addTypename>
        <VideoProjectFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );

    expect( videoProjectFiles.exists() ).toEqual( true );
    expect( videoProjectFiles.html() ).toEqual( null );
  } );

  it( 'does not render VideoProjectFile if unit.files is `null`', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullFilesMocks } addTypename>
        <VideoProjectFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );
    const videoProjectFile = videoProjectFiles.find( 'VideoProjectFile' );

    expect( videoProjectFiles.exists() ).toEqual( true );
    expect( videoProjectFile.exists() ).toEqual( false );
  } );

  it( 'does not render VideoProjectFile if unit.files is `[]`', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noFilesMocks }>
        <VideoProjectFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const videoProjectFiles = wrapper.find( 'VideoProjectFiles' );
    const videoProjectFile = videoProjectFiles.find( 'VideoProjectFile' );

    expect( videoProjectFiles.exists() ).toEqual( true );
    expect( videoProjectFile.exists() ).toEqual( false );
  } );
} );
