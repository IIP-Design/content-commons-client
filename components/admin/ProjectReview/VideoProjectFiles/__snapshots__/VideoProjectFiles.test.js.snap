// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoProjectFiles /> renders the final state 1`] = `
<VideoProjectFiles
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
/>
`;
